### ISSUE TITLE
Hidden types in module system type error

### ISSUE THREAD

**Issue by eponier** [coq/coq#6151]
#### Version
8.6.1 && master

#### Operating system
Ubuntu 16.04

#### Description of the problem
I was defining a small module and get a particularly useless message, where some types are replaced with `??`.
```
Module Type Sig.
  Parameter P : nat -> Prop.
  Parameter P_true : forall n, P n.
End Sig.

Module M <: Sig.
  Definition P (n:nat) := True.
  Definition P_true := true.
End M.
(* Signature components for label P_true do not match:
   expected type "??" but found type "bool". *)

Module M <: Sig.
  Definition P (n:nat) := True.
  Lemma P_true : forall (b:bool), P (if b then 1 else 0).
  Proof. intros. constructor. Qed.
End M.
(* Signature components for label P_true do not match:
   expected type "??" but found type "??". *)
```

This strange behavior did not exist in 8.5pl3.

**Comment by ejgallego** [comment]
Thanks @eponier , I guess the first step is to get a bisect of the bad commit.

**Comment by SkySkimmer** [comment]
Fixed since about 8.11
